cpu/stm32l4/periph/adc.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (C) 2014-2016 Freie Universität Berlin
+ * Copyright (C) 2018 HAW-Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     cpu_stm32l4
+ * @ingroup     drivers_periph_adc
+ * @{
+ *
+ * @file
+ * @brief       Low-level ADC driver implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Michel Rottleuthner <michel.rottleuthner@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+#include "xtimer.h"
+
+/**
+ * @brief map CPU specific register/value names
+ */
+#if defined(CPU_MODEL_STM32L476RG)
+#define ADC_CR_REG      CR
+#define ADC_ISR_REG     ISR
+#define ADC_PERIPH_CLK  AHB2
+/* on stm32-l476rg all ADC clocks are are enabled by this bit
+   further clock config is possible over CKMODE[1:0] bits in ADC_CCR reg */
+#define ADC_CLK_EN_MASK   (RCC_AHB2ENR_ADCEN)
+/* refering to Datasheet Section 6.3.18 (ADC characteristics) the minimum
+   achievable sampling rate is 4.21 Msps (12 Bit resolution on slow channel)
+   we use that worst case for configuring the sampling time to be sure it
+   works on all channels.
+   TCONV = Sampling time + 12.5 ADC clock cycles.
+   At 80MHz this means we need to set SMP to 001 (6.5 ADC clock cycles) to
+   stay within specs. (80000000/(6.5+12.5)) = 4210526   */
+#define ADC_SMP_MIN_VAL      (0x1)
+
+/* The sampling time can be specified for each channel over SMPR1 and SMPR2.
+   This specifies the first channel that goes to SMPR2 instead of SMPR1. */
+#define ADC_SMP_BIT_WIDTH    (3)
+
+/* The sampling time can be specified for each channel over SMPR1 and SMPR2.
+   This specifies the first channel that goes to SMPR2 instead of SMPR1. */
+#define ADC_SMPR2_FIRST_CHAN (10)
+#endif
+
+/**
+ * @brief   Load the ADC configuration
+ */
+static const adc_conf_t adc_config[] = ADC_CONFIG;
+
+/**
+ * @brief   Allocate locks for all three available ADC devices
+ */
+static mutex_t locks[] = {
```

Comment:
```static mutex_t locks[ADC_DEVS];``` will have the same effect.

Reply:
I allowed myself to directly address your comment @kaspar030. Tested with nucleo-l476 and still works.